[annotables]: https://github.com/stephenturner/annotables/
[AnnotationHub]: https://bioconductor.org/packages/AnnotationHub/
[AppVeyor CI]: https://ci.appveyor.com/
[assertive]: https://cran.r-project.org/package=assertive
[basejump]: https://basejump.acidgenomics.com/
[bcbio]: https://bcbio-nextgen.readthedocs.io/
[bcbioBase]: http://bioinformatics.sph.harvard.edu/bcbioBase/
[bcbioRNASeq]: https://bioinformatics.sph.harvard.edu/bcbioRNASeq/
[bcbioSingleCell]: http://bioinformatics.sph.harvard.edu/bcbioSingleCell/
[bcl2fastq]: https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html
[BiocInstaller]: https://bioconductor.org/packages/BiocInstaller/
[BiocManager]: https://cran.r-project.org/package=BiocManager
[Bioconductor]: https://bioconductor.org/
[Biostrings]: https://bioconductor.org/packages/Biostrings/
[brightworm]: https://brightworm.acidgenomics.com/
[Cell Ranger]: https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger
[clusterProfiler]: https://bioconductor.org/packages/clusterProfiler/
[conda]: https://conda.io/
[covr]: https://cran.r-project.org/package=covr
[DESeq2]: https://bioconductor.org/packages/DESeq2/
[devtools]: https://cran.r-project.org/package=devtools
[dplyr]: http://dplyr.tidyverse.org/
[edgeR]: https://bioconductor.org/packages/edgeR/
[EggNOG]: http://eggnogdb.embl.de/
[Ensembl]: https://www.ensembl.org/
[ensembldb]: https://bioconductor.org/packages/ensembldb/
[Entrez]: https://www.ncbi.nlm.nih.gov/search/
[Excel]: https://products.office.com/en-us/excel/
[fgsea]: http://bioconductor.org/packages/fgsea/
[GenomeInfoDb]: https://bioconductor.org/packages/GenomeInfoDb/
[GenomeInfoDbData]: https://bioconductor.org/packages/GenomeInfoDbData/
[ggplot2]: https://ggplot2.tidyverse.org/
[GitHub]: https://github.com/
[glue]: https://glue.tidyverse.org/
[GO]: http://www.geneontology.org/
[GRCh37]: https://grch37.ensembl.org/
[GSEA]: http://software.broadinstitute.org/gsea/
[HGNC]: https://www.genenames.org/
[HMS RC]: https://rc.hms.harvard.edu/
[inDrops]: http://1cell-bio.com/
[KEGG]: https://www.genome.jp/kegg/
[LibreOffice]: https://www.libreoffice.org/
[lintr]: https://github.com/jimhester/lintr/
[macOS]: https://www.apple.com/macos/
[magrittr]: https://magrittr.tidyverse.org/
[Markdown]: https://daringfireball.net/projects/markdown/
[Matrix]: https://cran.r-project.org/package=Matrix
[MGI]: http://www.informatics.jax.org/
[MSigDb]: http://software.broadinstitute.org/gsea/msigdb/
[NIH]: https://www.nih.gov/
[PANTHER]: http://pantherdb.org/
[pathview]: https://bioconductor.org/packages/pathview/
[pheatmap]: https://cran.r-project.org/package=pheatmap
[pointillism]: https://pointillism.acidgenomics.com/
[Python]: https://python.org/
[r-wormbase]: https://wormbase.acidgenomics.com/
[R]: https://www.r-project.org/
[rapmap]: https://github.com/COMBINE-lab/RapMap/
[readr]: http://readr.tidyverse.org/
[rio]: https://cran.r-project.org/package=rio
[rlang]: http://rlang.tidyverse.org/
[roxygen2]: https://cran.r-project.org/package=roxygen2
[RStudio]: https://www.rstudio.com/
[salmon]: https://combine-lab.github.io/salmon/
[scater]: https://bioconductor.org/packages/scater/
[scran]: https://bioconductor.org/packages/scran/
[scRNA-seq]: http://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html#single-cell-rna-sequencing
[Seurat]: http://satijalab.org/seurat/
[SummarizedExperiment]: https://bioconductor.org/packages/SummarizedExperiment/
[testthat]: https://cran.r-project.org/package=testthat
[tibble]: https://tibble.tidyverse.org/
[tidyr]: http://tidyr.tidyverse.org/
[tidyverse]: http://www.tidyverse.org/
[Travis CI]: https://travis-ci.com/
[tximport]: https://bioconductor.org/packages/tximport/
[UCSC]: https://genome.ucsc.edu/
[UMAP]: https://github.com/lmcinnes/umap/
[umis]: https://github.com/vals/umis/
[Windows]: https://www.microsoft.com/en-us/windows/
[WormBase]: http://www.wormbase.org/
[zinbwave]: https://bioconductor.org/packages/zinbwave/
[zingeR]: https://github.com/statOmics/zingeR/
