<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert Ensembl Transcripts to Genes — convertTranscriptsToGenes • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Convert Ensembl Transcripts to Genes — convertTranscriptsToGenes" />

<meta property="og:description" content="Convert Ensembl Transcripts to Genes" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert Ensembl Transcripts to Genes</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/steinbaugh/basejump/blob/master/R/convertTranscriptsToGenes-methods.R'><code>R/convertTranscriptsToGenes-methods.R</code></a></small>
    <div class="hidden name"><code>convertTranscriptsToGenes.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Convert Ensembl Transcripts to Genes</p>
    
    </div>

    <pre class="usage"><span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for character</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>build</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for matrix</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>build</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>aggregate</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># S4 method for sparseMatrix</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>,
  <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>build</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>aggregate</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Passthrough to <code><a href='deprecated.html'>makeTx2geneFromEnsembl()</a></code>.</p></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p><code>Tx2Gene</code> or <code>NULL</code>. Transcript-to-gene mappings. If set
<code>NULL</code>, the function will attempt to download the mappings from Ensembl
automatically.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code>. Full latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>build</th>
      <td><p><code>string</code> or <code>NULL</code>. Genome build assembly name (e.g.
"<code>GRCh38</code>"). If set <code>NULL</code>, defaults to the most recent build available.</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><code>scalar integer</code> or <code>NULL</code>. Release version (e.g. <code>90</code>). If
set <code>NULL</code>, defaults to the most recent release available.</p></td>
    </tr>
    <tr>
      <th>aggregate</th>
      <td><p><code>boolean</code>. For objects supporting <code>dim()</code>, aggregate counts
to gene level and collapse the matrix.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<ul>
<li><p><code>character</code>: <code>factor</code>. Genes in the values, transcripts in the names.</p></li>
<li><p><code>matrix</code>, <code>sparseMatrix</code>, <code>SummarizedExperiment</code>: Object containing counts
collapsed to gene level by default (see <code>aggregate</code> argument).</p></li>
</ul>

    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>For objects containing a count matrix, the object rows will be collapsed to
gene level using <code><a href='aggregate.html'>aggregateRows()</a></code>. This applies to our
<code>SummarizedExperiment</code> method.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='aggregate.html'>aggregateRows()</a></code>.</p>
<p>Other Annotation Functions: <code><a href='convertGenesToSymbols.html'>convertGenesToSymbols</a></code>,
  <code><a href='convertUCSCBuildToEnsembl.html'>convertUCSCBuildToEnsembl</a></code>,
  <code><a href='emptyRanges.html'>emptyRanges</a></code>, <code><a href='geneSynonyms.html'>geneSynonyms</a></code>,
  <code><a href='makeGRanges.html'>makeGRanges</a></code>, <code><a href='makeGene2Symbol.html'>makeGene2Symbol</a></code>,
  <code><a href='makeTx2Gene.html'>makeTx2Gene</a></code>, <code><a href='panther.html'>panther</a></code>,
  <code><a href='stripTranscriptVersions.html'>stripTranscriptVersions</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>tx_se_small</span>)
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>tx_se_small</span>

<span class='no'>tx2gene</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='tx2gene.html'>tx2gene</a></span>(<span class='no'>object</span>)
<span class='fu'>print</span>(<span class='no'>tx2gene</span>)</div><div class='output co'>#&gt; Tx2Gene with 6 rows and 2 columns
#&gt;                    transcriptID          geneID
#&gt;                     &lt;character&gt;     &lt;character&gt;
#&gt; ENST00000494424 ENST00000494424 ENSG00000000003
#&gt; ENST00000496771 ENST00000496771 ENSG00000000003
#&gt; ENST00000612152 ENST00000612152 ENSG00000000003
#&gt; ENST00000371584 ENST00000371584 ENSG00000000419
#&gt; ENST00000371588 ENST00000371588 ENSG00000000419
#&gt; ENST00000413082 ENST00000413082 ENSG00000000419</div><div class='input'><span class='no'>transcripts</span> <span class='kw'>&lt;-</span> <span class='fu'>rownames</span>(<span class='no'>object</span>)
<span class='fu'>print</span>(<span class='no'>transcripts</span>)</div><div class='output co'>#&gt; [1] "ENST00000494424" "ENST00000496771" "ENST00000612152" "ENST00000371584"
#&gt; [5] "ENST00000371588" "ENST00000413082"</div><div class='input'>
<span class='co'># character ====</span>
<span class='co'># Returns as factor.</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>transcripts</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='no'>tx2gene</span>)
<span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; ENST00000494424 ENST00000496771 ENST00000612152 ENST00000371584 ENST00000371588 
#&gt; ENSG00000000003 ENSG00000000003 ENSG00000000003 ENSG00000000419 ENSG00000000419 
#&gt; ENST00000413082 
#&gt; ENSG00000000419 
#&gt; Levels: ENSG00000000003 ENSG00000000419</div><div class='input'><span class='fu'>str</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;  Factor w/ 2 levels "ENSG00000000003",..: 1 1 1 2 2 2
#&gt;  - attr(*, "names")= chr [1:6] "ENST00000494424" "ENST00000496771" "ENST00000612152" "ENST00000371584" ...</div><div class='input'>
<span class='co'># matrix ====</span>
<span class='co'># Note that transcript IDs currently must be in the rows.</span>
<span class='no'>counts</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='counts.html'>counts</a></span>(<span class='no'>object</span>)
<span class='fu'>print</span>(<span class='no'>counts</span>)</div><div class='output co'>#&gt;                 sample1 sample2 sample3 sample4
#&gt; ENST00000494424       1       2       3       4
#&gt; ENST00000496771       5       6       7       8
#&gt; ENST00000612152       9      10      11      12
#&gt; ENST00000371584      13      14      15      16
#&gt; ENST00000371588      17      18      19      20
#&gt; ENST00000413082      21      22      23      24</div><div class='input'><span class='co'># Aggregate to gene level.</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>counts</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='no'>tx2gene</span>, <span class='kw'>aggregate</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; <span class='message'>Aggregating counts using `sum()`...</span>
#&gt; <span class='message'>Groupings:</span>
#&gt; <span class='message'>ENST00000494424 ENST00000496771 ENST00000612152 ENST00000371584 ENST00000371588 </span>
#&gt; <span class='message'>ENSG00000000003 ENSG00000000003 ENSG00000000003 ENSG00000000419 ENSG00000000419 </span>
#&gt; <span class='message'>ENST00000413082 </span>
#&gt; <span class='message'>ENSG00000000419 </span>
#&gt; <span class='message'>Levels: ENSG00000000003 ENSG00000000419</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;                 sample1 sample2 sample3 sample4
#&gt; ENSG00000000003      15      18      21      24
#&gt; ENSG00000000419      51      54      57      60</div><div class='input'><span class='fu'>colSums</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; sample1 sample2 sample3 sample4 
#&gt;      66      72      78      84 </div><div class='input'><span class='co'># Simply map to rownames.</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>counts</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='no'>tx2gene</span>, <span class='kw'>aggregate</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
<span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;                 sample1 sample2 sample3 sample4
#&gt; ENSG00000000003       1       2       3       4
#&gt; ENSG00000000003       5       6       7       8
#&gt; ENSG00000000003       9      10      11      12
#&gt; ENSG00000000419      13      14      15      16
#&gt; ENSG00000000419      17      18      19      20
#&gt; ENSG00000000419      21      22      23      24</div><div class='input'><span class='fu'>colSums</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; sample1 sample2 sample3 sample4 
#&gt;      66      72      78      84 </div><div class='input'>
<span class='co'># SummarizedExperiment ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; <span class='message'>Aggregating counts using `sum()`...</span>
#&gt; <span class='message'>Groupings:</span>
#&gt; <span class='message'>ENST00000494424 ENST00000496771 ENST00000612152 ENST00000371584 ENST00000371588 </span>
#&gt; <span class='message'>ENSG00000000003 ENSG00000000003 ENSG00000000003 ENSG00000000419 ENSG00000000419 </span>
#&gt; <span class='message'>ENST00000413082 </span>
#&gt; <span class='message'>ENSG00000000419 </span>
#&gt; <span class='message'>Levels: ENSG00000000003 ENSG00000000419</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SummarizedExperiment 
#&gt; dim: 2 4 
#&gt; metadata(0):
#&gt; assays(1): counts
#&gt; rownames(2): ENSG00000000003 ENSG00000000419
#&gt; rowData names(0):
#&gt; colnames(4): sample1 sample2 sample3 sample4
#&gt; colData names(0):</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

