<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generate Empty Genomic Ranges — emptyRanges • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Generate Empty Genomic Ranges — emptyRanges" />

<meta property="og:description" content="Utility function that provides support for creating internal GRanges for
transgene and FASTA spike-in sequences." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.6.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate Empty Genomic Ranges</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/emptyRanges.R'><code>R/emptyRanges.R</code></a></small>
    <div class="hidden name"><code>emptyRanges.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Utility function that provides support for creating internal <code>GRanges</code> for
transgene and FASTA spike-in sequences.</p>
    
    </div>

    <pre class="usage"><span class='fu'>emptyRanges</span>(<span class='no'>names</span>, <span class='kw'>seqname</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"unknown"</span>, <span class='st'>"transgene"</span>, <span class='st'>"spike"</span>),
  <span class='kw'>mcolsNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>names</th>
      <td><p><code>character</code>. Gene or transcript names.</p></td>
    </tr>
    <tr>
      <th>seqname</th>
      <td><p><code>string</code> Name of the alternative chromosome to be defined in
<code><a href='http://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo'>GenomeInfoDb::seqnames()</a></code> where these ranges will be grouped. Defaults to
"<code>unknown</code>" but "<code>transgene</code>" (transgenes) and "<code>spike</code>" (spike-ins) are
also supported.</p></td>
    </tr>
    <tr>
      <th>mcolsNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Metadata column names to be defined
in the <code><a href='http://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class'>S4Vectors::mcols()</a></code> of the <code>GRanges</code> return. Normally this does not
need to be defined; useful when combining with another <code>GRanges</code> that
contains metadata.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>GRanges</code>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code>help("seqinfo", "GenomeInfoDb")</code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/makeExampleDESeqDataSet'>DESeq2::makeExampleDESeqDataSet()</a></code>.</p></li>
</ul></div>
    <div class='dont-index'><p>Other Annotation Functions: <code><a href='annotable.html'>annotable</a></code>,
  <code><a href='broadClass.html'>broadClass</a></code>,
  <code><a href='convertGenesToSymbols.html'>convertGenesToSymbols</a></code>,
  <code><a href='convertTranscriptsToGenes.html'>convertTranscriptsToGenes</a></code>,
  <code><a href='convertUCSCBuildToEnsembl.html'>convertUCSCBuildToEnsembl</a></code>,
  <code><a href='detectOrganism.html'>detectOrganism</a></code>, <code><a href='eggnog.html'>eggnog</a></code>,
  <code><a href='geneSynonyms.html'>geneSynonyms</a></code>, <code><a href='hgnc2gene.html'>hgnc2gene</a></code>,
  <code><a href='makeGRangesFromEnsembl.html'>makeGRangesFromEnsembl</a></code>,
  <code><a href='makeGRangesFromGFF.html'>makeGRangesFromGFF</a></code>,
  <code><a href='makeGene2symbolFromEnsembl.html'>makeGene2symbolFromEnsembl</a></code>,
  <code><a href='makeGene2symbolFromGFF.html'>makeGene2symbolFromGFF</a></code>,
  <code><a href='makeTx2geneFromEnsembl.html'>makeTx2geneFromEnsembl</a></code>,
  <code><a href='makeTx2geneFromGFF.html'>makeTx2geneFromGFF</a></code>, <code><a href='mgi2gene.html'>mgi2gene</a></code>,
  <code><a href='panther.html'>panther</a></code>,
  <code><a href='stripTranscriptVersions.html'>stripTranscriptVersions</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Unknown/dead genes</span>
<span class='fu'>emptyRanges</span>(<span class='st'>"ENSG00000000000"</span>, <span class='kw'>seqname</span> <span class='kw'>=</span> <span class='st'>"unknown"</span>)</div><div class='output co'>#&gt; GRanges object with 1 range and 0 metadata columns:
#&gt;                   seqnames    ranges strand
#&gt;                      &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;   ENSG00000000000  unknown     1-100      *
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'># Transgenes</span>
<span class='fu'>emptyRanges</span>(<span class='fu'>c</span>(<span class='st'>"EGFP"</span>, <span class='st'>"TDTOMATO"</span>, <span class='st'>"GAL4"</span>), <span class='kw'>seqname</span> <span class='kw'>=</span> <span class='st'>"transgene"</span>)</div><div class='output co'>#&gt; GRanges object with 3 ranges and 0 metadata columns:
#&gt;             seqnames    ranges strand
#&gt;                &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;       EGFP transgene     1-100      *
#&gt;   TDTOMATO transgene   101-200      *
#&gt;       GAL4 transgene   201-300      *
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'># Spike-ins</span>
<span class='fu'>emptyRanges</span>(<span class='st'>"ERCC"</span>, <span class='kw'>seqname</span> <span class='kw'>=</span> <span class='st'>"spike"</span>)</div><div class='output co'>#&gt; GRanges object with 1 range and 0 metadata columns:
#&gt;        seqnames    ranges strand
#&gt;           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;   ERCC    spike     1-100      *
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

