<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-s-classes" class="hasAnchor"><a href="#section-s-classes" class="anchor"></a>S4 Classes</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="EggNOG-class.html">EggNOG-class</a></code> </p>
        </td>
        <td><p>EggNOG Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Ensembl2Entrez-class.html">Ensembl2Entrez-class</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Gene2Symbol-class.html">Gene2Symbol-class</a></code> </p>
        </td>
        <td><p>Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="HGNC2Ensembl-class.html">HGNC2Ensembl-class</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MGI2Ensembl-class.html">MGI2Ensembl-class</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PANTHER-class.html">PANTHER-class</a></code> </p>
        </td>
        <td><p>PANTHER Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Tx2Gene-class.html">Tx2Gene-class</a></code> </p>
        </td>
        <td><p>Transcript-to-Gene Identifier Mappings</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-s-generators" class="hasAnchor"><a href="#section-s-generators" class="anchor"></a>S4 Generators</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="eggnog.html">eggnog()</a></code> </p>
        </td>
        <td><p>EggNOG Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ensembl2entrez.html">ensembl2entrez()</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene2symbol.html">gene2symbol()</a></code> </p>
        </td>
        <td><p>Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hgnc2ensembl.html">hgnc2ensembl()</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mgi2ensembl.html">mgi2ensembl()</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="panther.html">panther()</a></code> </p>
        </td>
        <td><p>PANTHER Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tx2gene.html">tx2gene()</a></code> </p>
        </td>
        <td><p>Transcript-to-Gene Identifier Mappings</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-genome-annotations" class="hasAnchor"><a href="#section-genome-annotations" class="anchor"></a>Genome Annotations</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="makeGRanges.html">makeGRangesFromEnsembl()</a></code> <code><a href="makeGRanges.html">makeGRangesFromGFF()</a></code> <code><a href="makeGRanges.html">annotable()</a></code> </p>
        </td>
        <td><p>Make Genomic Ranges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGene2Symbol.html">makeGene2SymbolFromEnsembl()</a></code> <code><a href="makeGene2Symbol.html">makeGene2SymbolFromGFF()</a></code> </p>
        </td>
        <td><p>Make Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeTx2Gene.html">makeTx2GeneFromEnsembl()</a></code> <code><a href="makeTx2Gene.html">makeTx2GeneFromGFF()</a></code> </p>
        </td>
        <td><p>Make Transcript-to-Gene Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneSynonyms.html">geneSynonyms()</a></code> </p>
        </td>
        <td><p>Gene Synonyms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="eggnog.html">eggnog()</a></code> </p>
        </td>
        <td><p>EggNOG Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="panther.html">panther()</a></code> </p>
        </td>
        <td><p>PANTHER Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="emptyRanges.html">emptyRanges()</a></code> </p>
        </td>
        <td><p>Generate Empty Genomic Ranges</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-identifier-mapping" class="hasAnchor"><a href="#section-gene-identifier-mapping" class="anchor"></a>Gene Identifier Mapping</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="convertGenesToSymbols.html">convertGenesToSymbols()</a></code> <code><a href="convertGenesToSymbols.html">convertSymbolsToGenes()</a></code> </p>
        </td>
        <td><p>Convert Ensembl Identifiers to Gene Symbols</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSampleIDsToNames.html">convertSampleIDsToNames()</a></code> </p>
        </td>
        <td><p>Convert Sample Identifiers to Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertTranscriptsToGenes.html">convertTranscriptsToGenes()</a></code> </p>
        </td>
        <td><p>Convert Ensembl Transcripts to Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertUCSCBuildToEnsembl.html">convertUCSCBuildToEnsembl()</a></code> </p>
        </td>
        <td><p>Convert UCSC Build to Ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapGenes.html">mapGenesToRownames()</a></code> <code><a href="mapGenes.html">mapGenesToIDs()</a></code> <code><a href="mapGenes.html">mapGenesToSymbols()</a></code> </p>
        </td>
        <td><p>Map Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene2symbol.html">gene2symbol()</a></code> </p>
        </td>
        <td><p>Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tx2gene.html">tx2gene()</a></code> </p>
        </td>
        <td><p>Transcript-to-Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ensembl2entrez.html">ensembl2entrez()</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hgnc2ensembl.html">hgnc2ensembl()</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mgi2ensembl.html">mgi2ensembl()</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl Gene Identifier Mappings</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-generic-methods" class="hasAnchor"><a href="#section-generic-methods" class="anchor"></a>Generic Methods</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aggregate.html">aggregateRows()</a></code> <code><a href="aggregate.html">aggregateCols()</a></code> </p>
        </td>
        <td><p>Aggregate Rows or Columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="as.html">coerceS4ToList()</a></code> <code><a href="as.html">as.list(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Force an Object to Belong to a Class</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combine.html">combine(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="combine.html">combine(<i>&lt;SingleCellExperiment&gt;</i>,<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Combine Multiple Objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdown.html">markdown()</a></code> </p>
        </td>
        <td><p>Markdown</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism.html">organism(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;GRanges&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Organism Accessor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="show.html">show(<i>&lt;PANTHER&gt;</i>)</a></code> </p>
        </td>
        <td><p>Show an Object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-import-export" class="hasAnchor"><a href="#section-import-export" class="anchor"></a>Import/Export</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="assignAndSaveData.html">assignAndSaveData()</a></code> </p>
        </td>
        <td><p>Assign and Save Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="export.html">export()</a></code> </p>
        </td>
        <td><p>Export</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="import.html">import()</a></code> </p>
        </td>
        <td><p>Import</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadData.html">loadData()</a></code> </p>
        </td>
        <td><p>Load Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadDataAsName.html">loadDataAsName()</a></code> </p>
        </td>
        <td><p>Load Data as Name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadRemoteData.html">loadRemoteData()</a></code> </p>
        </td>
        <td><p>Load Remote Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveData.html">saveData()</a></code> </p>
        </td>
        <td><p>Save Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveRdExamples.html">saveRdExamples()</a></code> </p>
        </td>
        <td><p>Save R Documentation Examples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transmit.html">transmit()</a></code> </p>
        </td>
        <td><p>Transmit Files from a Remote FTP Server</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="writeCounts.html">writeCounts()</a></code> </p>
        </td>
        <td><p>Write Counts</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-metadata" class="hasAnchor"><a href="#section-metadata" class="anchor"></a>Metadata</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="detectHPC.html">detectHPC()</a></code> </p>
        </td>
        <td><p>Detect HPC Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectLanes.html">detectLanes()</a></code> </p>
        </td>
        <td><p>Detect Sequencing Lanes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectOrganism.html">detectOrganism()</a></code> </p>
        </td>
        <td><p>Detect Organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="interestingGroups.html">interestingGroups()</a></code> <code><a href="interestingGroups.html">`interestingGroups<-`()</a></code> </p>
        </td>
        <td><p>Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="microplate.html">microplate()</a></code> </p>
        </td>
        <td><p>Microtiter Plate Well Identifiers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="minimalSampleData.html">minimalSampleData()</a></code> </p>
        </td>
        <td><p>Minimal Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism.html">organism(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;GRanges&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Organism Accessor</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleData.html">sampleData()</a></code> <code><a href="sampleData.html">`sampleData<-`()</a></code> <code><a href="sampleData.html">sampleNames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleNames.html">`sampleNames<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="uniteInterestingGroups.html">uniteInterestingGroups()</a></code> </p>
        </td>
        <td><p>Unite Interesting Groups</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-summarizedexperiment" class="hasAnchor"><a href="#section-summarizedexperiment" class="anchor"></a>SummarizedExperiment</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="counts.html">counts(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="counts.html">`counts<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;ANY&gt;</i>)</a></code> <code><a href="counts.html">`counts<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;matrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="interestingGroups.html">interestingGroups()</a></code> <code><a href="interestingGroups.html">`interestingGroups<-`()</a></code> </p>
        </td>
        <td><p>Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSummarizedExperiment.html">makeSummarizedExperiment()</a></code> </p>
        </td>
        <td><p>Make Summarized Experiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapGenes.html">mapGenesToRownames()</a></code> <code><a href="mapGenes.html">mapGenesToIDs()</a></code> <code><a href="mapGenes.html">mapGenesToSymbols()</a></code> </p>
        </td>
        <td><p>Map Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meltCounts.html">meltCounts()</a></code> </p>
        </td>
        <td><p>Melt Count Matrix to Long Format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="metrics.html">metrics()</a></code> <code><a href="metrics.html">metricsPerSample()</a></code> </p>
        </td>
        <td><p>Metrics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="minimalSampleData.html">minimalSampleData()</a></code> </p>
        </td>
        <td><p>Minimal Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleData.html">sampleData()</a></code> <code><a href="sampleData.html">`sampleData<-`()</a></code> <code><a href="sampleData.html">sampleNames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleNames.html">`sampleNames<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="selectSamples.html">selectSamples()</a></code> </p>
        </td>
        <td><p>Select Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tpm.html">tpm(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Transcripts Per Million (TPM)</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-singlecellexperiment" class="hasAnchor"><a href="#section-singlecellexperiment" class="anchor"></a>SingleCellExperiment</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aggregateCellsToSamples.html">aggregateCellsToSamples()</a></code> </p>
        </td>
        <td><p>Aggregate Cells to Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cell2sample.html">cell2sample()</a></code> </p>
        </td>
        <td><p>Cell-to-Sample Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSingleCellExperiment.html">makeSingleCellExperiment()</a></code> </p>
        </td>
        <td><p>Make Single-Cell Experiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapCellsToSamples.html">mapCellsToSamples()</a></code> </p>
        </td>
        <td><p>Map Cells to Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetPerSample.html">subsetPerSample()</a></code> </p>
        </td>
        <td><p>Subset Per Sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="topCellsPerSample.html">topCellsPerSample()</a></code> </p>
        </td>
        <td><p>Top Cells per Sample</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plots" class="hasAnchor"><a href="#section-plots" class="anchor"></a>Plots</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="plotCountsPerBiotype.html">plotCountsPerBiotype()</a></code> <code><a href="plotCountsPerBiotype.html">plotCountsPerBroadClass()</a></code> </p>
        </td>
        <td><p>Plot Counts per Biotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerGene.html">plotCountsPerGene()</a></code> </p>
        </td>
        <td><p>Plot Counts Per Gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenderMarkers.html">plotGenderMarkers()</a></code> </p>
        </td>
        <td><p>Plot Sexually Dimorphic Gender Marker Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGene.html">plotGene()</a></code> </p>
        </td>
        <td><p>Plot Gene Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenesDetected.html">plotGenesDetected()</a></code> </p>
        </td>
        <td><p>Plot Genes Detected</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotHeatmap.html">plotHeatmap()</a></code> <code><a href="plotHeatmap.html">plotCorrelationHeatmap()</a></code> <code><a href="plotHeatmap.html">plotQuantileHeatmap()</a></code> </p>
        </td>
        <td><p>Plot Heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotQC.html">plotQC()</a></code> </p>
        </td>
        <td><p>Plot Quality Control</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTotalCounts.html">plotTotalCounts()</a></code> </p>
        </td>
        <td><p>Plot Total Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotZerosVsDepth.html">plotZerosVsDepth()</a></code> </p>
        </td>
        <td><p>Plot Percentage of Zeros vs. Library Depth</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-ggplot-" class="hasAnchor"><a href="#section-ggplot-" class="anchor"></a>ggplot2</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="ggplot2-geoms.html">basejump_geom_abline()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_average()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_repel()</a></code> </p>
        </td>
        <td><p>ggplot2 Geometric Objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ggplot2-themes.html">theme_paperwhite()</a></code> <code><a href="ggplot2-themes.html">theme_midnight()</a></code> </p>
        </td>
        <td><p>ggplot2 Themes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sanitization" class="hasAnchor"><a href="#section-sanitization" class="anchor"></a>Sanitization</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="collapseToString.html">collapseToString()</a></code> </p>
        </td>
        <td><p>Collapse to String</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeNames.html">camel()</a></code> <code><a href="makeNames.html">dotted()</a></code> <code><a href="makeNames.html">snake()</a></code> <code><a href="makeNames.html">upperCamel()</a></code> <code><a href="makeNames.html">makeNames()</a></code> <code><a href="makeNames.html">makeDimnames()</a></code> </p>
        </td>
        <td><p>Make Syntactically Valid Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeNA.html">removeNA()</a></code> </p>
        </td>
        <td><p>Remove Rows and Columns Containing Only <code>NA</code> Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeNA.html">sanitizeNA()</a></code> </p>
        </td>
        <td><p>Sanitize <code>NA</code> Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizePercent.html">sanitizePercent()</a></code> </p>
        </td>
        <td><p>Sanitize Percent</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeRowData.html">sanitizeRowData()</a></code> </p>
        </td>
        <td><p>Sanitize Row Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeSampleData.html">sanitizeSampleData()</a></code> </p>
        </td>
        <td><p>Sanitize Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sortUnique.html">sortUnique()</a></code> </p>
        </td>
        <td><p>Sort and Make Unique</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stripTranscriptVersions.html">stripTranscriptVersions()</a></code> </p>
        </td>
        <td><p>Strip Transcript Versions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toStringUnique.html">toStringUnique()</a></code> </p>
        </td>
        <td><p>Convert to a Unique Character String</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-calculations" class="hasAnchor"><a href="#section-calculations" class="anchor"></a>Calculations</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aggregate.html">aggregateRows()</a></code> <code><a href="aggregate.html">aggregateCols()</a></code> </p>
        </td>
        <td><p>Aggregate Rows or Columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geometricMean.html">geometricMean()</a></code> </p>
        </td>
        <td><p>Geometric Mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="logRatio.html">foldChangeToLogRatio()</a></code> <code><a href="logRatio.html">logRatioToFoldChange()</a></code> </p>
        </td>
        <td><p>Interconvert Log Ratio and Fold Change Values</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-markdown" class="hasAnchor"><a href="#section-markdown" class="anchor"></a>Markdown</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="markdown.html">markdown()</a></code> </p>
        </td>
        <td><p>Markdown</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownHeader.html">markdownHeader()</a></code> <code><a href="markdownHeader.html">mdHeader()</a></code> </p>
        </td>
        <td><p>Markdown Header</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownLink.html">markdownLink()</a></code> <code><a href="markdownLink.html">mdLink()</a></code> </p>
        </td>
        <td><p>Markdown Hyperlink</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownList.html">markdownList()</a></code> <code><a href="markdownList.html">mdList()</a></code> </p>
        </td>
        <td><p>Markdown List</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownPlotlist.html">markdownPlotlist()</a></code> <code><a href="markdownPlotlist.html">mdPlotlist()</a></code> </p>
        </td>
        <td><p>Markdown Plotlist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownTables.html">markdownTables()</a></code> </p>
        </td>
        <td><p>Create Multiple Tables in LaTeX, HTML, Markdown and reStructuredText</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareTemplate.html">prepareTemplate()</a></code> </p>
        </td>
        <td><p>Prepare R Markdown Template File</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-documentation" class="hasAnchor"><a href="#section-documentation" class="anchor"></a>Documentation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="RdTags.html">RdTags()</a></code> </p>
        </td>
        <td><p>R Documentation Tags</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseRd.html">parseRd()</a></code> </p>
        </td>
        <td><p>Parse R Documentation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveRdExamples.html">saveRdExamples()</a></code> </p>
        </td>
        <td><p>Save R Documentation Examples</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-assert-checks" class="hasAnchor"><a href="#section-assert-checks" class="anchor"></a>Assert Checks</h2>
          <p class="section-desc"><p>Readable check functions to ensure code integrity.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="assertAllAreNonExisting.html">assertAllAreNonExisting()</a></code> </p>
        </td>
        <td><p>Assert All Variables Are Non-Existing</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertAllAreURL.html">assertAllAreURL()</a></code> <code><a href="assertAllAreURL.html">isURL()</a></code> </p>
        </td>
        <td><p>Assert All Are URL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertAllAreUniqueGeneNames.html">assertAllAreUniqueGeneNames()</a></code> </p>
        </td>
        <td><p>Assert All Are Unique Gene Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertAllAreValidNames.html">assertAllAreValidNames()</a></code> <code><a href="assertAllAreValidNames.html">assertHasValidDimnames()</a></code> <code><a href="assertAllAreValidNames.html">assertHasValidNames()</a></code> <code><a href="assertAllAreValidNames.html">validDimnames()</a></code> <code><a href="assertAllAreValidNames.html">validNames()</a></code> </p>
        </td>
        <td><p>Assert All Are Valid Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertAreGeneAnnotations.html">assertAreGeneAnnotations()</a></code> </p>
        </td>
        <td><p>Assert Are Gene Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertAreTranscriptAnnotations.html">assertAreTranscriptAnnotations()</a></code> </p>
        </td>
        <td><p>Assert Are Ensembl Transcript Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertFormalCompress.html">assertFormalCompress()</a></code> </p>
        </td>
        <td><p>Assert Formal Compression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertFormalGene2Symbol.html">assertFormalGene2Symbol()</a></code> </p>
        </td>
        <td><p>Assert Formal Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertFormalInterestingGroups.html">assertFormalInterestingGroups()</a></code> </p>
        </td>
        <td><p>Interesting Groups Formal Assert Check</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertHasRownames.html">assertHasRownames()</a></code> <code><a href="assertHasRownames.html">hasRownames()</a></code> </p>
        </td>
        <td><p>Assert Has Rownames</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsANumberOrNULL.html">assertIsANumberOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is a Number or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsAStringOrNULL.html">assertIsAStringOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is a String or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsAlpha.html">assertIsAlpha()</a></code> </p>
        </td>
        <td><p>Assert Is Alpha Level</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsAnIntegerOrNULL.html">assertIsAnIntegerOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is an Integer or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsColorScaleContinuousOrNULL.html">assertIsColorScaleContinuousOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is Color Palette Scale Continuous or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsColorScaleDiscreteOrNULL.html">assertIsColorScaleDiscreteOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is Color Palette Scale Discrete or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsDir.html">assertIsDir()</a></code> </p>
        </td>
        <td><p>Assert Is Directory</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsFillScaleContinuousOrNULL.html">assertIsFillScaleContinuousOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is Fill Palette Scale Continuous or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsFillScaleDiscreteOrNULL.html">assertIsFillScaleDiscreteOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is Fill Palette Scale Discrete or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsHeaderLevel.html">assertIsHeaderLevel()</a></code> </p>
        </td>
        <td><p>Assert Is Markdown Header Level</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsHexColorFunctionOrNULL.html">assertIsHexColorFunctionOrNULL()</a></code> </p>
        </td>
        <td><p>Assert Is Hex Color Function or NULL</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertIsImplicitInteger.html">assertIsAnImplicitInteger()</a></code> <code><a href="assertIsImplicitInteger.html">assertIsAnImplicitIntegerOrNULL()</a></code> <code><a href="assertIsImplicitInteger.html">assertIsImplicitInteger()</a></code> <code><a href="assertIsImplicitInteger.html">assertIsImplicitIntegerOrNULL()</a></code> <code><a href="assertIsImplicitInteger.html">isAnImplicitInteger()</a></code> <code><a href="assertIsImplicitInteger.html">isImplicitInteger()</a></code> </p>
        </td>
        <td><p>Assert Is Implicit Integer</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-miscellaneous" class="hasAnchor"><a href="#section-miscellaneous" class="anchor"></a>Miscellaneous</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="checkClasses.html">checkClasses()</a></code> </p>
        </td>
        <td><p>Check Classes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectHPC.html">detectHPC()</a></code> </p>
        </td>
        <td><p>Detect HPC Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dots.html">dots()</a></code> </p>
        </td>
        <td><p>Extract Dots from Function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="headtail.html">headtail()</a></code> </p>
        </td>
        <td><p>Return the First and Last Part of an Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="initDir.html">initDir()</a></code> </p>
        </td>
        <td><p>Initialize Directory</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="localOrRemoteFile.html">localOrRemoteFile()</a></code> </p>
        </td>
        <td><p>Dynamically Handle a Local or Remote File Path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchArgsToDoCall.html">matchArgsToDoCall()</a></code> </p>
        </td>
        <td><p>Match Arguments to Do Call</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchInterestingGroups.html">matchInterestingGroups()</a></code> </p>
        </td>
        <td><p>Match Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MethodDefinition.html">methodFormals()</a></code> <code><a href="MethodDefinition.html">methodFunction()</a></code> </p>
        </td>
        <td><p>Access to the Data Inside an S4 Method Definition</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiassignAsEnvir.html">multiassignAsEnvir()</a></code> </p>
        </td>
        <td><p>Assign Multiple Objects as an Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="printString.html">printString()</a></code> </p>
        </td>
        <td><p>Print String</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="realpath.html">realpath()</a></code> </p>
        </td>
        <td><p>Express File Paths in Canonical Form</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="separator.html">separator()</a></code> </p>
        </td>
        <td><p>Separator Bar</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="showSlotInfo.html">showSlotInfo()</a></code> </p>
        </td>
        <td><p>Show Slot Information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="standardizeCall.html">standardizeCall()</a></code> </p>
        </td>
        <td><p>Standardize Call</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-example-data-sets" class="hasAnchor"><a href="#section-example-data-sets" class="anchor"></a>Example Data Sets</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="rse_small.html">rse_small</a></code> </p>
        </td>
        <td><p>Example RNA-Seq Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sce_small.html">sce_small</a></code> </p>
        </td>
        <td><p>Example Single-Cell RNA-Seq Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tx_se_small.html">tx_se_small</a></code> </p>
        </td>
        <td><p>Example Transcript-Level Data Set</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-s-classes">S4 Classes</a></li>
      <li><a href="#section-s-generators">S4 Generators</a></li>
      <li><a href="#section-genome-annotations">Genome Annotations</a></li>
      <li><a href="#section-gene-identifier-mapping">Gene Identifier Mapping</a></li>
      <li><a href="#section-generic-methods">Generic Methods</a></li>
      <li><a href="#section-import-export">Import/Export</a></li>
      <li><a href="#section-metadata">Metadata</a></li>
      <li><a href="#section-summarizedexperiment">SummarizedExperiment</a></li>
      <li><a href="#section-singlecellexperiment">SingleCellExperiment</a></li>
      <li><a href="#section-plots">Plots</a></li>
      <li><a href="#section-ggplot-">ggplot2</a></li>
      <li><a href="#section-sanitization">Sanitization</a></li>
      <li><a href="#section-calculations">Calculations</a></li>
      <li><a href="#section-markdown">Markdown</a></li>
      <li><a href="#section-documentation">Documentation</a></li>
      <li><a href="#section-assert-checks">Assert Checks</a></li>
      <li><a href="#section-miscellaneous">Miscellaneous</a></li>
      <li><a href="#section-example-data-sets">Example Data Sets</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

